### ISSUE TITLE
Usage as loss

### ISSUE THREAD

**Issue by cipri-tom** [githubharald/CTCWordBeamSearch#4]
Hi,

Thank you for such complete repo with code examples and frameworks in both Python and CPP and tensorflow ops. Really amazing work! 

If I understand correctly, this is only to be used during evaluation/prediction phase, to get more meaningful ones. I was wondering if it could also be used as a loss during training, such that paths which follow the language model are more probable than those which do not.

Or it is actually not relevant since CTC squashes all paths that produce the label and therefore all versions are maximised ?

Many thanks, again, for making your contributions public !

**Comment by githubharald** [comment]
Hi,

you are right, this decoder is used in the prediction phase to get better results by incorporating a language model / word dictionary.
The CTC loss takes the RNN output matrix and the ground truth text and tries all possible alignments (i.e. paths) of the ground truth text. For each alignment, it calculates the probability. Finally, the sum of all those probabilities is taken and represents the total probability of seeing the given ground truth text in the matrix.
Short introduction to CTC + some follow-up links: https://stats.stackexchange.com/questions/320868/understanding-connectionist-temporal-classification-ctc/331376#331376

Are you working on a HTR or ASR system or are just asking out of curiosity?

Regards
Harry

**Comment by cipri-tom** [comment]
Hi Harry,

Thank you for your reply and for the links! They confirm how I (weakly) understood CTC. Dunno why I imagined we can benefit from an external language model during training.  

--- 

I am working on HTR, yes, and at the moment using the `tf.nn.ctc_beam_search_decoder` version. I have to decode language-free strings and language-dependent strings, in the same task. The thing is the language-free ones can include language chars so I'm not sure the distinction made by WordBeamSearch would help me in its current phase, where it switches its `mode` based on the character. I believe I'll add a manual parameter to tell it in which mode to decode.

Final question: is Beam-Search decoding equivalent to char-level ngram ?

**Comment by githubharald** [comment]
Hi,

> Final question: is Beam-Search decoding equivalent to char-level ngram ?

char-level n-gram is a Language Model (LM) which assigns scores to consecutive groups of n chars, e.g. n=2 for bigrams. As an example, a char-group like "th" is more likely than "tx" in the English language.
Beam search decoding adds (at most) one single character at each iteration, therefore it is easy to integrate a char-level LM: each time a char is added, take the probability of the last char in the beam and the newly added char (in case of n=2) and multiply this textual score with the optical score. For a Python implementation see https://github.com/githubharald/CTCDecoder

**Comment by cipri-tom** [comment]
Thank you, Harry, for the detailed explanation!

Indeed, the question came from that other repo of yours, which has the mode 
> Beam Search Decoding: iteratively searches for best labeling, uses a character-level LM

So I was wondering whether this repo will also get that mode.

**Comment by cipri-tom** [comment]
hey, congratulations for the ICFHR'18 award ! ^^ 🎉

**Comment by githubharald** [comment]
thanks :-) Have you also been there?

**Comment by cipri-tom** [comment]
nope, but I checked the website for papers and posters and saw your prize on the front page 😃. I don't mean to spam your repo, so we can keep in touch by email

**Comment by githubharald** [comment]
cool ;-) Yeah, let's keep in touch by email, you can find my email in the file doc/poster.pdf
